lgc/patch/PatchWorkarounds.cpp

```diff
@@ -89,30 +89,48 @@ void PatchWorkarounds::applyImageDescWorkaround(void) {
   if (!m_pipelineState->getOptions().disableImageResourceCheck &&
       m_pipelineState->getTargetInfo().getGpuWorkarounds().gfx10.waFixBadImageDescriptor) {
 
+    // We have to consider waterfall.last.use as this may be used on a resource
+    // descriptor that is then used by an image instruction.
+    // waterfall.last.use needs to then be proceesed first due to the nature of
+    // the intrinsic (dst needs to go into the processed list to prevent it
+    // being processed twice, plus the workaround breaks the last.use if not
+    // handled like this)
+
+    // Build up a list of uses (since we modify we can't process them immediately)
+    SmallVector<CallInst *, 8> useWorkListLastUse;
+    SmallVector<CallInst *, 8> useWorkListImage;
+
     for (const Function &func : m_module->getFunctionList()) {
-      if (func.getName().startswith("llvm.amdgcn.image")) {
-        // Build up a list of uses (since we modify we can't process them immediately)
-        SmallVector<CallInst *, 8> useWorkList;
+      if (func.getName().startswith("llvm.amdgcn.waterfall.last.use") ||
+          func.getName().startswith("llvm.amdgcn.image")) {
         for (auto &use : func.uses()) {
           if (auto *callInst = dyn_cast<CallInst>(use.getUser())) {
             if (callInst->isCallee(&use))
-              useWorkList.push_back(callInst);
+              if (func.getName().startswith("llvm.amdgcn.waterfall.last.use"))
```

Comment:
I was going to say "can't we avoid the extra string comparison inside the loop by remembering which one we're on from above?". But, now I look at it, we should avoid all the string comparisons by using the intrinsic id instead. Would it make sense to do that in this change, since you're touching it anyway?

Reply:
Yes - in the case of the last.use intrinsic I think that makes sense. Obviously, the other one doesn't (since it isn't a single intrinsic we're looking for).